serialization/rp_ser.c

```diff
@@ -0,0 +1,366 @@
+	}
+
+	return rp_os_response_signal(rp);
+}
+
+static rp_err_t rp_ser_event_parse(struct rp_ser *rp, CborValue *it)
+{
+	uint8_t event;
+
+	if (!cbor_value_is_simple_type(it) ||
+	    cbor_value_get_simple_type(it, &event) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	if (cbor_value_advance_fixed(it) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	return event_parse(rp, event, it);
+}
+
+static rp_err_t rp_ser_cmd_parse(struct rp_ser *rp, CborValue *it)
+{
+	rp_err_t err;
+	uint8_t cmd;
+
+	if (!cbor_value_is_simple_type(it) ||
+	    cbor_value_get_simple_type(it, &cmd) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	if (cbor_value_advance_fixed(it) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	err = cmd_execute(rp, cmd, it);
+	if (err) {
+		return err;
+	}
+
+	RP_LOG_DBG("Received command 0x%02x", cmd);
+
+	return RP_SUCCESS;
+}
+
+static rp_err_t rp_ser_received_data_parse(struct rp_ser *rp,
+					   const uint8_t *data, size_t len)
+{
+	rp_err_t err;
+	CborParser parser;
+	CborValue value;
+	CborValue recursed;
+	uint8_t packet_type;
+
+	if (cbor_parser_init(data, len, 0, &parser, &value) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	if (!cbor_value_is_array(&value)) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	if (cbor_value_enter_container(&value, &recursed) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	/* Get BLE packet type. */
+	if (!cbor_value_is_simple_type(&recursed) ||
+	    cbor_value_get_simple_type(&recursed,
+				       &packet_type) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	cbor_value_advance_fixed(&recursed);
+
+	switch (packet_type) {
+	case RP_SER_PACKET_TYPE_CMD:
+		RP_LOG_DBG("Command received");
+		err = rp_ser_cmd_parse(rp, &recursed);
+
+		break;
+
+	case RP_SER_PACKET_TYPE_EVENT:
+		RP_LOG_DBG("Event received");
+		err = rp_ser_event_parse(rp, &recursed);
+
+		break;
+
+	case RP_SER_PACKET_TYPE_RSP:
+		RP_LOG_DBG("Response received");
+		err = rp_ser_response_parse(rp, &recursed);
+
+		break;
+
+	default:
+		RP_LOG_ERR("Unknown packet received");
+		return RP_ERROR_NOT_SUPPORTED;
+	}
+
+	if (err) {
+		return err;
+	}
+
+	/* Be sure that we unpacked all data from the array */
+	if (!cbor_value_at_end(&recursed)) {
+		RP_LOG_ERR("Received more data than expected");
+		return RP_ERROR_INTERNAL;
+	}
+
+	if (cbor_value_leave_container(&value, &recursed) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	return RP_SUCCESS;
+}
+
+static void transport_handler(struct rp_trans_endpoint *endpoint,
+			      const uint8_t *buf, size_t length)
+{
+	struct rp_ser *rp = RP_CONTAINER_OF(endpoint, struct rp_ser, endpoint);
+
+	rp_ser_received_data_parse(rp, buf, length);
+}
+
+rp_err_t rp_ser_cmd_send(struct rp_ser *rp, struct rp_ser_encoder *encoder,
+			 cmd_rsp_handler_t rsp)
+{
+	rp_err_t err;
+
+	if (!rp || !encoder) {
+		return RP_ERROR_NULL;
+	}
+
+	if (encoder->packet_size < 1) {
+		return RP_ERROR_INVALID_PARAM;
+	}
+
+	if (rp->rsp_handler) {
+		return RP_ERROR_BUSY;
```

Comment:
Kind of related to [my flow control comment in the other PR](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/2008#discussion_r395112258): It could be better to let the user define a wait time, and block on some semaphore or mutex for the specified amount of time. This return code is the equivalent of K_NO_WAIT for kernel apis

Reply:
The mutex is probably the best solution here but it required to create the OS abstaraction for it. I will wait for final transport solution from @doki-nordic and then we can decide which solution we should use here.